app/grandchallenge/hanging_protocols/static/js/hanging_protocols/hanging_protocol_visualization.js

```diff
@@ -0,0 +1,90 @@
+function insertAfter(newNode, existingNode) {
+    existingNode.parentNode.insertBefore(newNode, existingNode.nextSibling);
+}
+
+function removeAllChildNodes(parent) {
+    while (parent.firstChild) {
+        parent.removeChild(parent.firstChild);
+    }
+}
+
+function showOrHideVisualizationDiv(divId, jsonString) {
+    if (jsonString != '' && jsonString != 'null') {
+            document.getElementById(divId).style.display = 'block';
+        } else {
+            document.getElementById(divId).style.display = 'none';
+    }
+}
+
+function getGridDimensions(json) {
+    let allX = [];
+    let allY = [];
+    let allW = [];
+    let allH = [];
+
+    for (let i = 0; i < json.length; i++) {
+        json[i].x == undefined ? undefined : allX.push(json[i].x);
+        json[i].y == undefined ? undefined : allY.push(json[i].y);
+        json[i].w == undefined ? undefined : allW.push(json[i].w);
+        json[i].h == undefined ? undefined : allH.push(json[i].h);
+    }
+    let totalWidth = (allX.length != json.length | allW.length != json.length) ? json.length : Math.max(...allX.map(function(num, idx) {return num+allW[idx]}));
+    let totalHeight = (allY.length != json.length | allH.length != json.length) ? 1 : Math.max(...allY.map(function(num, idx) {return num+allH[idx]}));
+
+    return [totalHeight, totalWidth]
+}
+
+function createViewportDiv(divId, viewportNum, viewportSpec, totalHeight, totalWidth) {
+    let viewportDiv = document.createElement("div");
+    viewportDiv.setAttribute('id', viewportSpec.viewport_name);

```

Comment:
If `viewportSpec.viewport_name` is not (yet) defined, this will create a div with `id="undefined"`. So if there are multiple empty objects there will be multiple divs with `id="undefined"` but id should always be unique. Also it will show undefined as the name of the div. I think it would be nicer if either the div is not created when the `viewport_name`  field is not (yet) defined, or that it will show no title and for example a red background or something to indicate that it is an invalid viewport.

Reply:
Good point. I went with a mix of the two. The div only gets created when the `viewport_name` is not undefined, and if the provided `viewport_name` is not one of the possible labels, it gets a red background and a warning that a valid viewport name needs to be selected.